pyfr/integrators/std/steppers.py

```diff
@@ -17,18 +17,26 @@ class StdEulerStepper(BaseStdStepper):
     stepper_has_errest = False
     stepper_nregs = 2
     stepper_order = 1
+    stepper_has_variable_dt = False
 
     @property
     def _stepper_nfevals(self):
         return self.nsteps
 
     def step(self, t, dt):
         add, rhs = self._add, self.system.rhs
+        preproc, postproc = self.system.preproc, self.system.postproc
```

Comment:
As in a continuation of an existing graph?

Reply:
Yes, so hook into existing graphs.  There are probably a quite few different ways of going about it, so you'll want to experiment and see what ends up being cleanest.

As a first pass the integrators basically want to declare 'variants' of rhs.  So maybe adding a version of rhs which accepts an extra parameter/some flags and then using this to know what set of graphs should be used.  These arguments can carry through to the graph construction routine.

Reply:
I don't see how post-proc can be linked to the same graph as RHS

Reply:
In most cases post-proc is followed immediately by the pre-proc of the next stage hence they can be fused.  (There will be an end which can't, however.)

Reply:
Even on the linked ones, every kernel in pre-proc is dependent on post-proc completely finishing, so it doesn't seem like putting them on the same graph would really change much besides adding more code complexity

Reply:
Firstly, that's not entirely true in a mix-element setting.  There is no need for element type X to wait for the post-proc filtering of element type Y to finish before it can move onto its pre-proc step.

 Secondly, even when kernels run in serial graphs help to reduce launch overhead.

Thirdly, from my understanding of the code — and correct me if I am wrong — post-proc typically does the filtering whilst pre-proc then goes and computes the minimum entropy as its first step using the same data.  Thus, in the linked case these kernels really should be fused (or rather the filter should have an optional capability to dump the minimum entropy which can be used in the linked case to elide an explicit kernel; or just have it always write out the minimum entropy in the element as a freebie since it will almost always be used).  This will save compute and memory bandwidth in the back-to-back post-pre-proc scenario.

Reply:
Even in the mix-element case, pre-proc can't run until an element _and its neighbors_ are done filtering, so in general, element type X does need to wait for element type Y to finish post-proc before it can run pre-proc. 

In the filter kernel, it's possible to change entmin to inout and set it after the filter stage, but the kernel in pre-proc that computes entmin also copies that value to the fpts, so it still needs to run or those two functions need to be separated and then we'd need to add some logic about when the pre-proc graphs need to do that and when not. At best, we're cutting out one pointwise kernel and one read-in of nupts scalars, and this is with the assumption that all time schemes are going to immediately pre-proc the solution that was just post-proc'd which may not be the case in the future. In the current implementation, the pre-processing and post-processing routines are pretty cleanly split, so unless there are noticeable overall performance differences, I think the benefits of keeping the code less convoluted outweigh it.

Reply:
Point taken on the pre-proc side.  However, the design still looks a little suspect.

In the common case why can the filter not compute the minimum over the fpts (at the start), then do the filtering, then compute the new entropy, its minimum, and write it out to the fpts?  Would this not eliminate one of the kernels entirely?  Do we use the minimum anywhere else?  Then the case for doing an rhs are if `uin` is filtered or not, and if the minimum entropy is available or not.  These arguments can easily be passed to the rhs graph function which can then decide what sets of kernels are needed.  The optional functionality should come down to a Mako if statement in the filter to decide what to do; automatic argument pruning will help with the rest.  Of course, an ability to filter a solution stand-alone will be needed but that is unavoidable.

The current pre-proc graph is non-ideal in that not only may re-read in data that has immediately been processed by post-proc, but also does MPI transfers with very little overlap.  The only kernel is `iintcent` which is pretty inexpensive.  This is really bad for scaling.  If my reading is correct, we force the proper minimum entropy to be available at the time rhs is called, even though rhs has no use for it.  Conversely, if everything was merged into the normal right hand side graphs we'd have much more flexibility about when the exchange happens and where other ancillary kernels run (at least most of the time).

Reply:
Yeah it can be done that way if we're going with the route that the RHS is always immediately called on the filtered solution. We'd still need the kernel to calculate the minimum entropy for the first RHS action.

Reply:
Yes, I think that's fine; or at least it is probably not worth trying to optimize away at this point.